{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e4f2fe41",
   "metadata": {},
   "source": [
    "# SP+ End-to-end Evaluation: Annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a5f7d6e",
   "metadata": {},
   "source": [
    "## 0. Setting up your environment\n",
    "### DONE!  This env is now available in terminal and also here as a ipynb kernel!\n",
    "### STEP 0 = select spp(python 3.9.6) as the kernel here!!!\n",
    "\n",
    "### Install spp client to your local device: \n",
    "`pip install -i https://pip.s2.allenai.org/simple spp_client==0.3.1`\n",
    "\n",
    "### make sure you have the other important notebook pieces\n",
    "`pip install layoutparser`  \n",
    "`pip install matplotlib`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aa22c98a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Should have MMDA version >= 0.3.5\n",
      "Name: mmda\n",
      "Version: 0.4.3\n",
      "Summary: MMDA - multimodal document analysis\n",
      "Home-page: None\n",
      "Author: None\n",
      "Author-email: Allen Institute for Artificial Intelligence <contact@allenai.org>\n",
      "License: Apache-2.0\n",
      "Location: /Users/carolinew/.pyenv/versions/3.9.6/envs/spp/lib/python3.9/site-packages\n",
      "Requires: necessary, tqdm, pandas, pdf2image, ncls, pydantic, pdfplumber, requests\n",
      "Required-by: spp-client\n"
     ]
    }
   ],
   "source": [
    "# STEP 1 = SANITY CHECKS! \n",
    "# TODO: comment these out after confirming you have the correct versions\n",
    "print(\"Should have MMDA version >= 0.3.5\")\n",
    "! pip show mmda\n",
    "\n",
    "# TODO: comment these out after confirming you have the correct versions\n",
    "print(\"Should have spp-client version == 0.3.1\")\n",
    "! pip show spp-client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1d45b565",
   "metadata": {},
   "outputs": [],
   "source": [
    "# STEP 2 = SETUP\n",
    "import json, csv\n",
    "from spp_client.client import (\n",
    "    mmda_span_group_to_text_via_dwp_words,\n",
    "    RelationTypeId,\n",
    "    SppClient,\n",
    ")\n",
    "client = SppClient.mk(env=\"dev\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8daa8ab5",
   "metadata": {},
   "source": [
    "## 1. Get data from annotation store via spp_client"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18b9d55b",
   "metadata": {},
   "source": [
    "### 1a. Get and set sources\n",
    "Retreive information about the sources for each annotation - mentions, bib-entries, bib-entry-parses, and pdf plumber"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b6b1dc26",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "####################################################################\n",
    "# ANNOTATOR TODO: change this to sha you're analyzing\n",
    "####################################################################\n",
    "\n",
    "#This sha will get overwritten when running autogenerater script\n",
    "sha = '2b7d176acede6925b28e40f7962b83296940aa07'\n",
    "\n",
    "\n",
    "sha = '496a7f1f2eae30d375fd5fb2c55829fcffc93dfa'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc696479",
   "metadata": {},
   "source": [
    "### 1b. Retrieve Annotations\n",
    "Using the sources obtained from the previous cell, we will now retreive the annotations for each source via spp_client."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "29cd21cc",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sha' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 14\u001b[0m\n\u001b[1;32m     10\u001b[0m spp_grobid_version \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdag-0.0.7-grobid-bib-detector-eval\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     11\u001b[0m spp_mmda_version \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdag-0.0.15-mmda\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     13\u001b[0m spp_grobid_doc, spp_grobid_relations \u001b[38;5;241m=\u001b[39m client\u001b[38;5;241m.\u001b[39mget_document(\n\u001b[0;32m---> 14\u001b[0m             \u001b[43msha\u001b[49m,\n\u001b[1;32m     15\u001b[0m             spp_grobid_version,\n\u001b[1;32m     16\u001b[0m             annotation_types\u001b[38;5;241m=\u001b[39m[\n\u001b[1;32m     17\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpages\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     18\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtokens\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     19\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrows\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     20\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbib_entries\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     21\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcitation_mentions\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     22\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbib_entry_titles\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     23\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbib_entry_dois\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     24\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbib_entry_years\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     25\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwords\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     26\u001b[0m             ],\n\u001b[1;32m     27\u001b[0m             relation_types\u001b[38;5;241m=\u001b[39m[\n\u001b[1;32m     28\u001b[0m                 RelationTypeId(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrefers_to\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcitation_mentions\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbib_entries\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     29\u001b[0m             ],\n\u001b[1;32m     30\u001b[0m             allow_missing\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m     31\u001b[0m             include_images\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m     32\u001b[0m         )\n\u001b[1;32m     34\u001b[0m spp_mmda_doc, spp_mmda_relations \u001b[38;5;241m=\u001b[39m client\u001b[38;5;241m.\u001b[39mget_document(\n\u001b[1;32m     35\u001b[0m             sha,\n\u001b[1;32m     36\u001b[0m             spp_mmda_version,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     52\u001b[0m             include_images\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m     53\u001b[0m         )\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sha' is not defined"
     ]
    }
   ],
   "source": [
    "# Get annotations and data as text using spp client\n",
    "from spp_client.client import (\n",
    "    mmda_span_group_to_text_via_dwp_words,\n",
    "    RelationTypeId,\n",
    "    SppClient,\n",
    ")\n",
    "client = SppClient.mk(env=\"dev\")\n",
    "\n",
    "# SPP versions for eval samples\n",
    "spp_grobid_version = \"dag-0.0.7-grobid-bib-detector-eval\"\n",
    "spp_mmda_version = \"dag-0.0.15-mmda\"\n",
    "\n",
    "spp_grobid_doc, spp_grobid_relations = client.get_document(\n",
    "            sha,\n",
    "            spp_grobid_version,\n",
    "            annotation_types=[\n",
    "                \"pages\",\n",
    "                \"tokens\",\n",
    "                \"rows\",\n",
    "                \"bib_entries\",\n",
    "                \"citation_mentions\",\n",
    "                \"bib_entry_titles\",\n",
    "                \"bib_entry_dois\",\n",
    "                \"bib_entry_years\",\n",
    "                \"words\"\n",
    "            ],\n",
    "            relation_types=[\n",
    "                RelationTypeId(\"refers_to\", \"citation_mentions\", \"bib_entries\")\n",
    "            ],\n",
    "            allow_missing=True,\n",
    "            include_images=True\n",
    "        )\n",
    "\n",
    "spp_mmda_doc, spp_mmda_relations = client.get_document(\n",
    "            sha,\n",
    "            spp_mmda_version,\n",
    "            annotation_types=[\n",
    "                \"pages\",\n",
    "                \"tokens\",\n",
    "                \"rows\",\n",
    "                \"bib_entries\",\n",
    "                \"citation_mentions\",\n",
    "                \"bib_entry_titles\",\n",
    "                \"bib_entry_dois\",\n",
    "                \"bib_entry_years\",\n",
    "                \"words\"\n",
    "            ],\n",
    "            relation_types=[\n",
    "                RelationTypeId(\"refers_to\", \"citation_mentions\", \"bib_entries\")\n",
    "            ],\n",
    "            allow_missing=True,\n",
    "            include_images=True\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "16dfacc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spp_grobid_doc.fields ['bib_entry_years', 'citation_mentions', 'bib_entry_titles', 'words', 'tokens', 'rows', 'bib_entry_dois', 'bib_entries', 'pages']\n",
      "spp_mmda_doc.fields   ['bib_entry_years', 'citation_mentions', 'bib_entry_titles', 'words', 'tokens', 'rows', 'bib_entry_dois', 'bib_entries', 'pages']\n"
     ]
    }
   ],
   "source": [
    "print('spp_grobid_doc.fields', spp_grobid_doc.fields)\n",
    "print('spp_mmda_doc.fields  ', spp_mmda_doc.fields)\n",
    "\n",
    "if 'bib_entries' not in spp_grobid_doc.fields:\n",
    "    print(\"No bib_entries in spp_grobid_doc!\")\n",
    "\n",
    "if 'bib_entries' not in spp_mmda_doc.fields:\n",
    "    print(\"No bib_entries in spp_mmda_doc!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670ccd6f",
   "metadata": {},
   "source": [
    "## 2. Visalize Mentions (draw boxes)\n",
    "Use draw_boxes to visualize each mention."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "726a378c",
   "metadata": {},
   "source": [
    "### 2a. Make Layout Rectangles for mentions so that vizes can be made for them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "699b78cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#imports\n",
    "from collections import defaultdict\n",
    "import layoutparser.elements as lpe\n",
    "import layoutparser.visualization as lpv\n",
    "from typing import List\n",
    "\n",
    "#function to create box groups from mentions on the mmda doc\n",
    "def make_boxgroups_mentions(mmda_doc):\n",
    "\n",
    "    result_boxes_by_page = defaultdict(list)\n",
    "    original_boxes_by_page = defaultdict(list)\n",
    "    \n",
    "    for mention in mmda_doc.citation_mentions:\n",
    "        for box in mention.box_group.boxes:\n",
    "            original_boxes_by_page[box.page].append(box) \n",
    "            \n",
    "    #mapping between pages and spans on page\n",
    "    for page, list_of_boxes in original_boxes_by_page.items():\n",
    "        page_w, page_h = mmda_doc.images[page].size\n",
    "\n",
    "\n",
    "        for box in list_of_boxes:\n",
    "            if box.page == page:\n",
    "                result_boxes_by_page[page].append(lpe.Rectangle(\n",
    "                        box.l * page_w,\n",
    "                        box.t * page_h,\n",
    "                        (box.l + box.w) * page_w,\n",
    "                        (box.t + box.h) * page_h\n",
    "                    ))\n",
    "            else:\n",
    "                raise ValueError('Page number')\n",
    "    return result_boxes_by_page"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40eaf7c9",
   "metadata": {},
   "source": [
    "## 3. Get references and s2 id's for each mention."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b34d8d19",
   "metadata": {},
   "source": [
    "### 3a. Retreive citation mention numbers and which bib entries they are linked to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "e390bde1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ie: for citation_mention '0', the bib_entry_id it's linked to is '3'. \n",
    "\n",
    "def make_citation_links_dict(document_relations):\n",
    "    # citation_mention ids to bib_entry ids\n",
    "    citation_links = defaultdict(int)\n",
    "\n",
    "    for relation in document_relations[list(document_relations.keys())[0]]:\n",
    "        citation_links[relation.from_id] = relation.to_id\n",
    "        \n",
    "    return dict(citation_links)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "0134e973",
   "metadata": {},
   "outputs": [],
   "source": [
    "spp_grobid_links = make_citation_links_dict(spp_grobid_relations)\n",
    "spp_mmda_links = make_citation_links_dict(spp_mmda_relations)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cb6d7f1",
   "metadata": {},
   "source": [
    "### 3b. Use API call to find what s2 paper each bib entry is linked to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "d82b4657",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_index_of_bib_entry_with_id(bib_entry_id, mmda_doc):\n",
    "    linked_bib_entry_index_list = [i for i, bib_entry in enumerate(mmda_doc.bib_entries) if bib_entry.id == bib_entry_id]\n",
    "    assert len(linked_bib_entry_index_list) == 1, f\"there should be exactly one bib_entry with the id {bib_entry_id}.\"\n",
    "    return linked_bib_entry_index_list[0]\n",
    "\n",
    "def extract_first_group_symbols(span_groups):\n",
    "    if len(span_groups) > 1:\n",
    "        print(\"warning: multiple span_groups detected when 0 or 1 expected\")\n",
    "        for i, sg in enumerate(span_groups):\n",
    "             print(i, sg.text)\n",
    "    try: \n",
    "        first_one = span_groups[0] \n",
    "        text = mmda_span_group_to_text_via_dwp_words(first_one)\n",
    "    except IndexError: # case where was no annotation detected so span_groups was empty list\n",
    "            text = \"\" # it can be empty or say \"no text detected\" or something\n",
    "    return text\n",
    "\n",
    "def extract_first_group_symbols_no_dwp(span_groups):\n",
    "    if len(span_groups) > 1:\n",
    "        print(\"warning: multiple span_groups detected when 0 or 1 expected\")\n",
    "        for i, sg in enumerate(span_groups):\n",
    "            print(i, sg.text)\n",
    "    try: \n",
    "        first_one = span_groups[0] \n",
    "        text = mmda_span_group_to_text_via_dwp_words(first_one)\n",
    "    except IndexError: # case where was no annotation detected so span_groups was empty list\n",
    "            text = \"\" # it can be empty or say \"no text detected\" or something\n",
    "    return text\n",
    "\n",
    "\n",
    "def get_bib_entry_title(citation_mention_id, mmda_doc, citation_links):\n",
    "    linked_bib_id = citation_links[citation_mention_id]\n",
    "    # print(\"linked_bib_id: \", linked_bib_id)\n",
    "    # Accessing the linked bib_entry this way is error-prone because MMDA annotations can't be assumed to be sorted by SpanGroup IDs\n",
    "    # bib_title_span_groups = mmda_doc.bib_entries[linked_bib_id].bib_entry_titles \n",
    "    # We have to find the correct bib_entry this way:\n",
    "    linked_bib_index = get_index_of_bib_entry_with_id(linked_bib_id, mmda_doc)\n",
    "    if len(mmda_doc.bib_entry_titles) == 0:\n",
    "        return None\n",
    "    bib_title_span_groups = mmda_doc.bib_entries[linked_bib_index].bib_entry_titles \n",
    "    text_words = extract_first_group_symbols(bib_title_span_groups)\n",
    "    if text_words == \"\":\n",
    "         text_words = None\n",
    "\n",
    "    return text_words\n",
    "\n",
    "def get_bib_entry_year(citation_mention_id, mmda_doc, citation_links):\n",
    "    linked_bib_id = citation_links[citation_mention_id]\n",
    "    linked_bib_index = get_index_of_bib_entry_with_id(linked_bib_id, mmda_doc)\n",
    "    if len(mmda_doc.bib_entry_years) == 0:\n",
    "        return None\n",
    "    bib_year_span_groups = mmda_doc.bib_entries[linked_bib_index].bib_entry_years \n",
    "    text_symbols = extract_first_group_symbols_no_dwp(bib_year_span_groups)\n",
    "    try:\n",
    "        year = int(text_symbols)\n",
    "    except:\n",
    "        # If the year fails to be mapped to integer, default to None\n",
    "        year = None\n",
    "\n",
    "    return year\n",
    "\n",
    "def get_bib_entry_doi(citation_mention_id, mmda_doc, citation_links):\n",
    "    linked_bib_id = citation_links[citation_mention_id]\n",
    "    linked_bib_index = get_index_of_bib_entry_with_id(linked_bib_id, mmda_doc)\n",
    "    if len(mmda_doc.bib_entry_dois) == 0: \n",
    "         return None\n",
    "    bib_doi_span_groups = mmda_doc.bib_entries[linked_bib_index].bib_entry_dois\n",
    "    text_symbols = extract_first_group_symbols_no_dwp(bib_doi_span_groups)\n",
    "    if text_symbols == \"\":\n",
    "         text_symbols = None\n",
    "    return text_symbols\n",
    "\n",
    "def get_full_bib_entry_text(citation_mention_id, mmda_doc, citation_links):\n",
    "    linked_bib_id = citation_links[citation_mention_id]\n",
    "    linked_bib_index = get_index_of_bib_entry_with_id(linked_bib_id, mmda_doc)\n",
    "\n",
    "    bib_span_groups = [mmda_doc.bib_entries[linked_bib_index]]\n",
    "    text_symbols = extract_first_group_symbols(bib_span_groups)\n",
    "\n",
    "    return text_symbols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "69523ad6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make api call to get linked papers\n",
    "import json\n",
    "import requests\n",
    "\n",
    "def make_payload(mention_id, mmda_doc, citation_links):\n",
    "    payload = dict()\n",
    "    # make payload similar to how s2matchmaker does\n",
    "    # https://github.com/allenai/s2matchmaker/blob/a88bc1415cc28c0e4c644e5394529229db0a9c13/s2matchmaker/model/bib_entry_info.py#L32\n",
    "    title = get_bib_entry_title(mention_id, mmda_doc, citation_links) # make sure exists -- if no title then no payload\n",
    "    if not title:\n",
    "        return payload\n",
    "    # optional attributes will have None value in the payload even it is missing from the attributes\n",
    "    year = get_bib_entry_year(mention_id, mmda_doc, citation_links) # make sure is int, no dwp\n",
    "    doi = get_bib_entry_doi(mention_id, mmda_doc, citation_links) # no dwp \n",
    "    \n",
    "    payload.update({'title': title, 'year': year, 'matchedDoi': doi})\n",
    "\n",
    "    return payload\n",
    "\n",
    "\n",
    "def bibLinkingApiPost(payload, end_point='http://pipeline-api.prod.s2.allenai.org/citation/match'):\n",
    "    payload_str = json.dumps([payload])\n",
    "    headers = {\n",
    "        'Content-Type': 'application/json'\n",
    "    }\n",
    "    response = requests.request(\"POST\", end_point, headers=headers, data=payload_str).json()\n",
    "    return response\n",
    "\n",
    "\n",
    "\n",
    "# get s2_ids linked to each mention\n",
    "def get_s2_ids(mmda_doc, citation_links):\n",
    "    s2_ids = []\n",
    "    for citation_mention in mmda_doc.citation_mentions:\n",
    "        if citation_mention.id not in citation_links:\n",
    "            s2_ids.append('NOT LINKED')\n",
    "            continue\n",
    "        payload = make_payload(citation_mention.id, mmda_doc, citation_links)\n",
    "        if not payload:\n",
    "            s2_ids.append(\"\")\n",
    "            continue\n",
    "        matched_s2_id = str((bibLinkingApiPost(payload))).replace('[', '').replace(']', '')\n",
    "        s2_ids.append(matched_s2_id)\n",
    "    return s2_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59efc1db",
   "metadata": {},
   "source": [
    "## 4. Annotation Info!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "7cd13912",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "page to annotate: 2\n"
     ]
    }
   ],
   "source": [
    "####################################################################\n",
    "# ANNOTATOR TODO: Record page # from output in annotation spreadsheet\n",
    "####################################################################\n",
    "\n",
    "# get random page number to review and annotate\n",
    "import random\n",
    "\n",
    "pages_w_mentions = [citation_mention.pages[0].id for citation_mention in spp_grobid_doc.citation_mentions]\n",
    "\n",
    "annot_page = random.choice(pages_w_mentions)\n",
    "print(\"page to annotate:\", annot_page)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "40b53886",
   "metadata": {},
   "outputs": [],
   "source": [
    "# code to give table of mentions and their linked bib_entry ids \n",
    "# (if the doc is annotated with bib_entries), and s2 linked papers \n",
    "import pandas as pd\n",
    "from IPython.display import HTML\n",
    "\n",
    "def generate_ref_info(mmda_doc, citation_links):\n",
    "    if 'bib_entries' in mmda_doc.fields:\n",
    "        s2_ids = get_s2_ids(mmda_doc, citation_links)\n",
    "        ref_info = {\n",
    "            # \"Mention ID\": [citation_mention.id for citation_mention in mmda_doc.citation_mentions],\n",
    "            \"Page\": [citation_mention.pages[0].id for citation_mention in mmda_doc.citation_mentions],\n",
    "            \"Mention text\": [citation_mention.symbols for citation_mention in mmda_doc.citation_mentions],\n",
    "            \"Linked Bib id\": [citation_links[citation_mention.id] if citation_mention.id in citation_links else \"NOT LINKED\" \n",
    "                              for citation_mention in mmda_doc.citation_mentions],\n",
    "            # Bib entry and title texts are DWP \"words\".texts\n",
    "            \"Linked Bib Entry Extracted Text\": [get_full_bib_entry_text(citation_mention.id, mmda_doc, citation_links)  if citation_mention.id in citation_links else \"NOT LINKED\"\n",
    "                                                for citation_mention in mmda_doc.citation_mentions],    \n",
    "            \"Linked Bib Entry Title\": [get_bib_entry_title(citation_mention.id, mmda_doc, citation_links) if citation_mention.id in citation_links else \"NOT LINKED\"\n",
    "                                       for citation_mention in mmda_doc.citation_mentions],    \n",
    "            \"S2 id:\": [ids if ids != str(-1) else \" \" for ids in s2_ids],\n",
    "            \"S2 url\": [f\"http://api.semanticscholar.org/corpusid:{ids}\" if ids != str(-1) else \" \" for ids in s2_ids]\n",
    "            }\n",
    "    else:\n",
    "        ref_info = {\n",
    "            # \"Mention ID\": [citation_mention.id for citation_mention in mmda_doc.citation_mentions],\n",
    "            \"Page\": [citation_mention.pages[0].id for citation_mention in mmda_doc.citation_mentions],\n",
    "            \"Mention text\": [citation_mention.symbols for citation_mention in mmda_doc.citation_mentions],\n",
    "            \"Linked Bib Entry Title\": [\"no bibs\" for citation_mention in mmda_doc.citation_mentions],    \n",
    "            \"Linked Bib Entry Extracted Text\": [\"no bibs\" for citation_mention in mmda_doc.citation_mentions],    \n",
    "            \"S2 id:\": [\"no bibs\" for citation_mention in mmda_doc.citation_mentions],\n",
    "            \"S2 url\": [\"no bibs\" for citation_mention in mmda_doc.citation_mentions]\n",
    "            }\n",
    "    return ref_info\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7dcc81c",
   "metadata": {},
   "source": [
    "## SPP-Grobid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "af7df437",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Page</th>\n",
       "      <th>Mention text</th>\n",
       "      <th>Linked Bib id</th>\n",
       "      <th>Linked Bib Entry Extracted Text</th>\n",
       "      <th>Linked Bib Entry Title</th>\n",
       "      <th>S2 id:</th>\n",
       "      <th>S2 url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>9</td>\n",
       "      <td>Analysis in Liquid Biopsy Identifies Circulating PCTAIRE-1 mRNA as Biomarker in NSCLC. Cancer Genomics Proteomics 2020, 17(1):91-100. Wijerathne H, Witek MA, Baird AE, Soper SA: Liquid</td>\n",
       "      <td>Analysis in Liquid Biopsy Identifies Circulating PCTAIRE-1 mRNA as Biomarker in NSCLC</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>9</td>\n",
       "      <td>Analysis in Liquid Biopsy Identifies Circulating PCTAIRE-1 mRNA as Biomarker in NSCLC. Cancer Genomics Proteomics 2020, 17(1):91-100. Wijerathne H, Witek MA, Baird AE, Soper SA: Liquid</td>\n",
       "      <td>Analysis in Liquid Biopsy Identifies Circulating PCTAIRE-1 mRNA as Biomarker in NSCLC</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[2]</td>\n",
       "      <td>28</td>\n",
       "      <td>30. Graubert A, Aguet F, Ravi A, Ardlie KG, Getz : RNA-SeQC 2: Efficient RNA-seq</td>\n",
       "      <td>RNA-SeQC 2: Efficient RNA-seq</td>\n",
       "      <td>232142793</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:232142793\" target=\"_blank\">http://api.semanticscholar.org/corpusid:232142793</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>[3]</td>\n",
       "      <td>12</td>\n",
       "      <td>IM, de Mol P, Smilde TJ et al: Liquid biopsy reveals KLK3 mRNA as a prognostic marker for progression free survival in patients with metastatic castration-resistant prostate cancer undergoing first-line abiraterone acetate and prednisone treatment. Mol Oncol 2021, 15(9):2453-2465. Morimoto O, Nagano</td>\n",
       "      <td>Liquid biopsy reveals KLK3 mRNA as a prognostic marker for progression free survival in patients with metastatic castration-resistant prostate cancer undergoing first-line abiraterone acetate and prednisone treatment</td>\n",
       "      <td>232090123</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:232090123\" target=\"_blank\">http://api.semanticscholar.org/corpusid:232090123</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>[4]</td>\n",
       "      <td>49</td>\n",
       "      <td>Methods 2017, 49(4):1494-1502. Tallarida RJ, Murray RB: Dunnett’s</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:\" target=\"_blank\">http://api.semanticscholar.org/corpusid:</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>[6]</td>\n",
       "      <td>31</td>\n",
       "      <td>specific mRNA integrity large-scale RNA-sequencing data . Nat Commun 2015, 6:7816. Al-Fageeh</td>\n",
       "      <td>specific mRNA integrity large-scale RNA-sequencing data</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>9</td>\n",
       "      <td>Analysis in Liquid Biopsy Identifies Circulating PCTAIRE-1 mRNA as Biomarker in NSCLC. Cancer Genomics Proteomics 2020, 17(1):91-100. Wijerathne H, Witek MA, Baird AE, Soper SA: Liquid</td>\n",
       "      <td>Analysis in Liquid Biopsy Identifies Circulating PCTAIRE-1 mRNA as Biomarker in NSCLC</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>[7]</td>\n",
       "      <td>45</td>\n",
       "      <td>assigning sequence reads genomic features . Bioinformatics 2014, 30 (7):923-930. Student: The probable error of a mean. Biometrika 1908, 6 :1-25.</td>\n",
       "      <td>assigning sequence reads genomic features</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>[8]</td>\n",
       "      <td>8</td>\n",
       "      <td>as liquid biopsy predicts chemo-sensitivity in advanced gastric cancer patients. J Cancer 2017, 8(3):434-442. Chang JW, CL, Wang CL</td>\n",
       "      <td>as liquid biopsy predicts chemo-sensitivity in advanced gastric cancer patients</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>[9]</td>\n",
       "      <td>43</td>\n",
       "      <td>BAM files . Source Code Biol Med 2014, 9 (1). Langmead B, Salzberg SL: Fast gapped-read</td>\n",
       "      <td>BAM files</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>[10]</td>\n",
       "      <td>2</td>\n",
       "      <td>Individuality and variation in gene expression patterns in human blood . P Natl Acad Sci USA 2003, 100 (4):1896-1901. Garrett-Bakelman FE, Darshi</td>\n",
       "      <td>Individuality and variation in gene expression patterns in human blood</td>\n",
       "      <td>7261038</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:7261038\" target=\"_blank\">http://api.semanticscholar.org/corpusid:7261038</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>[11]</td>\n",
       "      <td>57</td>\n",
       "      <td>accounting for selection bias. Genome Biol 2010, 11(2):R14. Timmons JA, Szkop KJ, Gallagher IJ: Multiple sources of</td>\n",
       "      <td>accounting for selection bias</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>[12]</td>\n",
       "      <td>56</td>\n",
       "      <td>extended review and a software tool . One 2017, 12 (12):e0190152. Young MD, Wakefield MJ, Smyth GK, Oshlack A: Gene ontology analysis</td>\n",
       "      <td>extended review and a software tool</td>\n",
       "      <td>31126451</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:31126451\" target=\"_blank\">http://api.semanticscholar.org/corpusid:31126451</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>[13]</td>\n",
       "      <td>6</td>\n",
       "      <td>risk assessment. Trends Mol Med 2007, 13(10):422-432. Sole C, Arnaiz E, Manterola L, Otaegui , Lawrie CH: The</td>\n",
       "      <td>risk assessment</td>\n",
       "      <td>1886659</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:1886659\" target=\"_blank\">http://api.semanticscholar.org/corpusid:1886659</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>[14]</td>\n",
       "      <td>57</td>\n",
       "      <td>accounting for selection bias. Genome Biol 2010, 11(2):R14. Timmons JA, Szkop KJ, Gallagher IJ: Multiple sources of</td>\n",
       "      <td>accounting for selection bias</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2</td>\n",
       "      <td>[15]</td>\n",
       "      <td>27</td>\n",
       "      <td>RNA-seq data with DESeq2. Genome Biol 2014, 15 :550. Monahan JF: Numerical methods of statistics : Cambridge</td>\n",
       "      <td>RNA-seq data with DESeq2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2</td>\n",
       "      <td>[16]</td>\n",
       "      <td>58</td>\n",
       "      <td>enrichment analysis of global -omics data . Genome Biol 2015, 16 :186. Rhee SY, Wood V, Dolinski K, Draghici S: Use and misuse of the</td>\n",
       "      <td>enrichment analysis of global -omics data</td>\n",
       "      <td>3212934</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:3212934\" target=\"_blank\">http://api.semanticscholar.org/corpusid:3212934</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2</td>\n",
       "      <td>[17]</td>\n",
       "      <td>60</td>\n",
       "      <td>and Pitfalls. Brief 2016, 17(5):808-818. Holm S: A simple sequentially rejective multiple</td>\n",
       "      <td>and</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2</td>\n",
       "      <td>[18]</td>\n",
       "      <td>24</td>\n",
       "      <td>heifers collected at the time of artificial insemination. Data Brief 2018, 18:706-709.</td>\n",
       "      <td>heifers collected at the time of artificial insemination</td>\n",
       "      <td>48363356</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:48363356\" target=\"_blank\">http://api.semanticscholar.org/corpusid:48363356</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2</td>\n",
       "      <td>[19]</td>\n",
       "      <td>28</td>\n",
       "      <td>30. Graubert A, Aguet F, Ravi A, Ardlie KG, Getz : RNA-SeQC 2: Efficient RNA-seq</td>\n",
       "      <td>RNA-SeQC 2: Efficient RNA-seq</td>\n",
       "      <td>232142793</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:232142793\" target=\"_blank\">http://api.semanticscholar.org/corpusid:232142793</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2</td>\n",
       "      <td>[20]</td>\n",
       "      <td>10</td>\n",
       "      <td>diagnosis. Expert Rev Mol Diagn 2020, 20(8):771-788. Modlin IM, Kidd M, Oberg K, Falconi M, Filosso PL,</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>208818363</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:208818363\" target=\"_blank\">http://api.semanticscholar.org/corpusid:208818363</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2</td>\n",
       "      <td>[21]</td>\n",
       "      <td>46</td>\n",
       "      <td>48. Kalpić D, Hlupić N, Lovrić M: Student’s t-Tests . In: International</td>\n",
       "      <td>Student’s t-Tests</td>\n",
       "      <td>35737096</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:35737096\" target=\"_blank\">http://api.semanticscholar.org/corpusid:35737096</a></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_info = generate_ref_info(spp_grobid_doc, spp_grobid_links)\n",
    "\n",
    "df = pd.DataFrame.from_dict(ref_info, orient='index')\n",
    "df = df.transpose()\n",
    "\n",
    "# selecting rows based on random page selected to annotate\n",
    "# you can comment the line below out to see all mentions across all pages\n",
    "df = df[df['Page'] == annot_page]\n",
    "\n",
    "# for bib viz below\n",
    "relevant_grobid_bibs = set(df['Linked Bib id'].tolist())\n",
    "\n",
    "\n",
    "# make s2 links clickable\n",
    "HTML(df.to_html(render_links=True, escape=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf4d7e70",
   "metadata": {},
   "source": [
    "## SPP-MMDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "db74a982",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Page</th>\n",
       "      <th>Mention text</th>\n",
       "      <th>Linked Bib id</th>\n",
       "      <th>Linked Bib Entry Extracted Text</th>\n",
       "      <th>Linked Bib Entry Title</th>\n",
       "      <th>S2 id:</th>\n",
       "      <th>S2 url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[2]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>[3]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>[4]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>[6]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>[1]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>[7]</td>\n",
       "      <td>2</td>\n",
       "      <td>Additional file 7. Gene ontology enrichment analysis of biological processes for the 666 genes with greater transcript abundance in PWBCs cryopreserved after 24 hours of 667 refrigeration post collection versus one hour within collection. 668</td>\n",
       "      <td>Additional file 7. Gene ontology enrichment analysis of biological processes for the 666 genes with greater transcript abundance in PWBCs cryopreserved after 24 hours of 667 refrigeration post collection</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>[8]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>[9]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2</td>\n",
       "      <td>[10]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2</td>\n",
       "      <td>[11]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>[12]</td>\n",
       "      <td>3</td>\n",
       "      <td>Additional file 12. Code utilized for the processing and analysis of the raw data and 688 analytical procedures employed in this study. 689</td>\n",
       "      <td>Additional file 12. Code utilized for the processing and analysis of the raw data and 688 analytical procedures employed in this study</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>[13]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>[14]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2</td>\n",
       "      <td>[15]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2</td>\n",
       "      <td>[16]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2</td>\n",
       "      <td>[17]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2</td>\n",
       "      <td>[18]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2</td>\n",
       "      <td>[19]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2</td>\n",
       "      <td>[20]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2</td>\n",
       "      <td>[21]</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td>NOT LINKED</td>\n",
       "      <td><a href=\"http://api.semanticscholar.org/corpusid:NOT LINKED\" target=\"_blank\">http://api.semanticscholar.org/corpusid:NOT LINKED</a></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_info = generate_ref_info(spp_mmda_doc, spp_mmda_links)\n",
    "\n",
    "df = pd.DataFrame.from_dict(ref_info, orient='index')\n",
    "df = df.transpose()\n",
    "\n",
    "# selecting rows based on random page selected to annotate\n",
    "# you can comment the line below out to see all mentions across all pages\n",
    "df = df[df['Page'] == annot_page]\n",
    "\n",
    "# for bib viz below\n",
    "relevant_mmda_bibs = set(df['Linked Bib id'].tolist())\n",
    "\n",
    "\n",
    "# make s2 links clickable\n",
    "HTML(df.to_html(render_links=True, escape=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efc17903",
   "metadata": {},
   "outputs": [],
   "source": [
    "#egor version\n",
    "def draw_boxgroups(image, rectangle_list, color_map=None, id_map=None):\n",
    "    viz = []\n",
    "    if color_map is None:\n",
    "        color_map={'color_1': 'blue', 'color_2': 'green', 'color_3': 'yellow', 'color_4': 'pink'}\n",
    "\n",
    "    for rectangle in rectangle_list:\n",
    "        viz.append(lpe.TextBlock(rectangle,\n",
    "                                 type=f'color_1'\n",
    "                                )\n",
    "                  )\n",
    "\n",
    "    if image:\n",
    "        return lpv.draw_box(image,\n",
    "                            viz,\n",
    "                            color_map=color_map,\n",
    "                            show_element_id=False,\n",
    "                            id_font_size=10)\n",
    "    \n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "745fc1e4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# mentions imgs\n",
    "import cv2\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "\n",
    "rows = 1\n",
    "columns = 2\n",
    "fontsize = 48\n",
    "\n",
    "print(\"annotation page: \", annot_page)\n",
    "\n",
    "# spp-grobid\n",
    "#make boxes and save as image\n",
    "boxes_by_page = make_boxgroups_mentions(spp_grobid_doc)\n",
    "image1 = draw_boxgroups(spp_grobid_doc.images[annot_page], boxes_by_page[annot_page])\n",
    "\n",
    "#plot\n",
    "fig = plt.figure(figsize=(50, 50))\n",
    "fig.add_subplot(rows, columns, 1)\n",
    "plt.axis('off')\n",
    "plt.title(\"SPP-Grobid Mentions\",fontdict={'fontsize': fontsize})\n",
    "plt.imshow(image1)\n",
    "\n",
    "\n",
    "#plot side by side\n",
    "fig.add_subplot(rows, columns, 2)\n",
    "plt.axis('off')\n",
    "plt.title(\"SPP-MMDA Mentions\",fontdict={'fontsize': fontsize})\n",
    "plt.imshow(image1)\n",
    "plt.subplots_adjust(bottom=0.15, wspace=0.05)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7423b66f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# View all mentions\n",
    "# uncomment below to see all pages that have mentions (not just the selected annotation page)\n",
    "\n",
    "# print(\"annotation page: \", annot_page)\n",
    "\n",
    "# for page in boxes_by_page:\n",
    "#     print(f\"page {page}:\")\n",
    "#     display(draw_boxgroups(spp_grobid_doc.images[page], boxes_by_page[page]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daa08795",
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_text_blocks(page, text_blocks_dictionary, color_map=None, id_map=None):\n",
    "    image = spp_grobid_doc.images[page]\n",
    "\n",
    "    if color_map is None:\n",
    "        color_map={'relevant': 'blue', 'other': '#f7ede2'}\n",
    "\n",
    "    if image:\n",
    "        return lpv.draw_box(image,\n",
    "                            text_blocks_dictionary[page],\n",
    "                            color_map=color_map,\n",
    "                            show_element_id=True,\n",
    "                            id_font_size=10,\n",
    "                            id_text_color='blue'\n",
    "                            )\n",
    "    \n",
    "    return None\n",
    "\n",
    "#function to create labeled text blocks from bibs on the mmda doc\n",
    "def make_text_blocks_bibs(mmda_doc, relevant_bib_entry_ids):\n",
    "\n",
    "    result_text_blocks_by_page = defaultdict(list)\n",
    "    original_boxes_by_page = defaultdict(list)\n",
    "    page_sizes = dict()\n",
    "    \n",
    "    for bib in mmda_doc.bib_entries:\n",
    "        for box in bib.box_group.boxes:\n",
    "            if box.page not in page_sizes:\n",
    "                page_sizes[box.page] = mmda_doc.images[box.page].size\n",
    "            page_w, page_h = page_sizes[box.page]\n",
    "\n",
    "            color = 'relevant' if bib.id in relevant_bib_entry_ids else 'other'\n",
    "\n",
    "            rect = lpe.Rectangle(\n",
    "                    box.l * page_w,\n",
    "                    box.t * page_h,\n",
    "                    (box.l + box.w) * page_w,\n",
    "                    (box.t + box.h) * page_h\n",
    "                )\n",
    "            result_text_blocks_by_page[box.page].append(lpe.TextBlock(\n",
    "                rect,\n",
    "                id=bib.id,\n",
    "                type=color\n",
    "            ))\n",
    "    return result_text_blocks_by_page\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2c008dc",
   "metadata": {},
   "source": [
    "### Bibs Pages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e0fc2db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bibs imgs\n",
    "spp_grobid_bib_pages = set([bib_entry.box_group.boxes[0].page for bib_entry in spp_grobid_doc.bib_entries])\n",
    "spp_mmda_bib_pages = set([bib_entry.box_group.boxes[0].page for bib_entry in spp_mmda_doc.bib_entries])\n",
    "print('spp_grobid_bib_pages', sorted(list(spp_grobid_bib_pages)))\n",
    "print('spp_mmda_bib_pages  ', sorted(list(spp_mmda_bib_pages)))\n",
    "\n",
    "bib_pages = spp_grobid_bib_pages.union(spp_mmda_bib_pages)\n",
    "bib_pages = sorted(list(bib_pages))\n",
    "print('bib_pages           ', bib_pages)\n",
    "\n",
    "\n",
    "rows = 1\n",
    "columns = 2\n",
    "fontsize = 48\n",
    "\n",
    "\n",
    "spp_grobid_bib_text_blocks = make_text_blocks_bibs(spp_grobid_doc, relevant_grobid_bibs)\n",
    "spp_mmda_bib_text_blocks = make_text_blocks_bibs(spp_mmda_doc, relevant_mmda_bibs)\n",
    "\n",
    "\n",
    "for i, page in enumerate(bib_pages):\n",
    "\n",
    "    #plot\n",
    "    fig = plt.figure(figsize=(50, 50))\n",
    "    fig.add_subplot(rows, columns, 1)\n",
    "    plt.axis('off')\n",
    "    if i == 0:\n",
    "        plt.title(\"SPP-Grobid Bibs\",fontdict={'fontsize': fontsize})\n",
    "    plt.imshow(draw_text_blocks(page, spp_grobid_bib_text_blocks)) \n",
    "\n",
    "\n",
    "    #plot side by side\n",
    "    fig.add_subplot(rows, columns, 2)\n",
    "    plt.axis('off')\n",
    "    if i == 0:\n",
    "        plt.title(\"SPP-MMDA Bibs\",fontdict={'fontsize': fontsize})\n",
    "    plt.imshow(draw_text_blocks(page, spp_mmda_bib_text_blocks)) \n",
    "    plt.subplots_adjust(bottom=0.15, wspace=0.05)\n",
    "\n",
    "print(\"COLOR LEGEND: Blue is the relevant bib (that the selected annotation page has mentions linking to) and pale yellow is irrelevant, just other bibs that the detector detected\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aae60d6e",
   "metadata": {},
   "source": [
    "## ANNOTATOR TODO: \n",
    "Record each mention on this page (regardless of if it got recognized by SPP or not) as 1 row in the spreadsheet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2151b987",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1347867e",
   "metadata": {},
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "cannot import name 'ModelField' from 'pydantic.fields' (/Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/pydantic/fields.py)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [15], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Get annotations and data as text using spp client\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mspp_client\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mclient\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[1;32m      3\u001b[0m     mmda_span_group_to_text_via_dwp_words,\n\u001b[1;32m      4\u001b[0m     RelationTypeId,\n\u001b[1;32m      5\u001b[0m     SppClient,\n\u001b[1;32m      6\u001b[0m )\n\u001b[1;32m      7\u001b[0m client \u001b[38;5;241m=\u001b[39m SppClient\u001b[38;5;241m.\u001b[39mmk(env\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdev\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m# http://spp-api.dev.s2.allenai.org/v1/document/496a7f1f2eae30d375fd5fb2c55829fcffc93dfa/0.0.22?annotationType=words&annotationType=term_contexts&annotationType=term_mentions&annotationType=bib_entries&includeAsAttrs=false\u001b[39;00m\n",
      "File \u001b[0;32m~/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py:19\u001b[0m\n\u001b[1;32m     <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py?line=15'>16</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mPIL\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mImage\u001b[39;00m \u001b[39mimport\u001b[39;00m Image \u001b[39mas\u001b[39;00m PILImage\n\u001b[1;32m     <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py?line=16'>17</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mrequests\u001b[39;00m\n\u001b[0;32m---> <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py?line=18'>19</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mai2_internal\u001b[39;00m \u001b[39mimport\u001b[39;00m api\n\u001b[1;32m     <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py?line=19'>20</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mannotation_store\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mclient\u001b[39;00m \u001b[39mimport\u001b[39;00m ApiClient\n\u001b[1;32m     <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py?line=20'>21</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mannotation_store\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mpdf_annotations\u001b[39;00m \u001b[39mimport\u001b[39;00m (\n\u001b[1;32m     <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py?line=21'>22</a>\u001b[0m     Box \u001b[39mas\u001b[39;00m AsBox,\n\u001b[1;32m     <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py?line=22'>23</a>\u001b[0m     PdfAnnotations \u001b[39mas\u001b[39;00m AsPdfAnnotations,\n\u001b[1;32m     <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/spp_client/client.py?line=23'>24</a>\u001b[0m )\n",
      "File \u001b[0;32m~/.pyenv/versions/3.9.6/lib/python3.9/site-packages/ai2_internal/api.py:4\u001b[0m\n\u001b[1;32m      <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/ai2_internal/api.py?line=0'>1</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mtyping\u001b[39;00m \u001b[39mimport\u001b[39;00m List, Optional, Type\n\u001b[1;32m      <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/ai2_internal/api.py?line=2'>3</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpydantic\u001b[39;00m \u001b[39mimport\u001b[39;00m BaseModel, Extra, Field\n\u001b[0;32m----> <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/ai2_internal/api.py?line=3'>4</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpydantic\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mfields\u001b[39;00m \u001b[39mimport\u001b[39;00m ModelField\n\u001b[1;32m      <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/ai2_internal/api.py?line=5'>6</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mmmda\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mtypes\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mannotation\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mmmda_ann\u001b[39;00m\n\u001b[1;32m      <a href='file:///Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/ai2_internal/api.py?line=7'>8</a>\u001b[0m __all__ \u001b[39m=\u001b[39m [\u001b[39m\"\u001b[39m\u001b[39mBoxGroup\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mSpanGroup\u001b[39m\u001b[39m\"\u001b[39m]\n",
      "\u001b[0;31mImportError\u001b[0m: cannot import name 'ModelField' from 'pydantic.fields' (/Users/carolinew/.pyenv/versions/3.9.6/lib/python3.9/site-packages/pydantic/fields.py)"
     ]
    }
   ],
   "source": [
    "# Get annotations and data as text using spp client\n",
    "from spp_client.client import (\n",
    "    mmda_span_group_to_text_via_dwp_words,\n",
    "    RelationTypeId,\n",
    "    SppClient,\n",
    ")\n",
    "client = SppClient.mk(env=\"dev\")\n",
    "\n",
    "# http://spp-api.dev.s2.allenai.org/v1/document/496a7f1f2eae30d375fd5fb2c55829fcffc93dfa/0.0.22?annotationType=words&annotationType=term_contexts&annotationType=term_mentions&annotationType=bib_entries&includeAsAttrs=false\n",
    "\n",
    "sha = '496a7f1f2eae30d375fd5fb2c55829fcffc93dfa'\n",
    "spp_mmda_version = \"0.0.22\"\n",
    "\n",
    "spp_mmda_doc, spp_mmda_relations = client.get_document(\n",
    "            sha,\n",
    "            spp_mmda_version,\n",
    "            annotation_types=[\n",
    "                \"pages\",\n",
    "                \"tokens\",\n",
    "                \"rows\",\n",
    "                \"bib_entries\",\n",
    "                \"term_contexts\",\n",
    "                \"term_mentions\"\n",
    "                \"words\"\n",
    "            ],\n",
    "            relation_types=[],\n",
    "            allow_missing=True,\n",
    "            include_images=True\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "43a5d5ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9a9fff7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['symbols', 'metadata', 'term_contexts', 'tokens', 'rows', 'bib_entries', 'pages'])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "interpreter": {
   "hash": "c3d423d9951799dc36cf6fbf27c9169c5ac24f9f16d3113630166130e0a716b1"
  },
  "kernelspec": {
   "display_name": "Python 3.9.6 64-bit ('spp')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "85aef6625182614c82ba3fada19dd46c3306619fa9c96361bbc593929a8b39ac"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
